File: ptp-haskell/test/runner/TestRunner.hs
module TestRunner
  ( Test
  , Assertion
  , testCase
  , testGroup
  , assertEqual
  , runTest
  ) where

type Assertion = IO (Maybe String)

data Test
  = Case String Assertion
  | Group String [Test]

testCase :: String -> Assertion -> Test
testCase = Case

testGroup :: String -> [Test] -> Test
testGroup = Group

assertEqual :: (Eq a, Show a) => a -> a -> Assertion
assertEqual expected actual
  | expected == actual = pure Nothing
  | otherwise = pure (Just message)
  where
    message = "expected " ++ show expected ++ ", but got " ++ show actual

runTest :: Test -> IO Int
runTest = runWithIndent 0

runWithIndent :: Int -> Test -> IO Int
runWithIndent level test =
  case test of
    Group name tests -> runGroup level name tests
    Case name assertion -> runCase level name assertion

runGroup :: Int -> String -> [Test] -> IO Int
runGroup level name tests = do
  putStrLn (renderGroupHeader level name)
  failures <- mapM (runWithIndent (level + 2)) tests
  putStrLn ""
  pure (sum failures)

runCase :: Int -> String -> Assertion -> IO Int
runCase level name assertion = do
  result <- assertion
  putStrLn (renderCaseLine level name result)
  pure (failureCount result)

renderGroupHeader :: Int -> String -> String
renderGroupHeader level name = indent level ++ name

renderCaseLine :: Int -> String -> Maybe String -> String
renderCaseLine level name result =
  indent level ++ casePrefix result ++ name ++ caseSuffix result

casePrefix :: Maybe String -> String
casePrefix Nothing = "[PASS] "
casePrefix (Just _) = "[FAIL] "

caseSuffix :: Maybe String -> String
caseSuffix Nothing = ""
caseSuffix (Just message) = " -> " ++ message

failureCount :: Maybe String -> Int
failureCount Nothing = 0
failureCount (Just _) = 1

indent :: Int -> String
indent level = replicate level ' '

File: ptp-prolog/test/runner/test_runner.pl
:- module(test_runner, [run_test/2]).

run_test(Test, Failures) :-
    run_test(Test, 0, Failures).

run_test(test_group(Name, Tests), Indent, Failures) :-
    write_indented_line(Indent, Name),
    NextIndent is Indent + 2,
    run_tests(Tests, NextIndent, Failures),
    nl.
run_test(test_case(Name, Expected, Actual, Goal), Indent, Failures) :-
    (   once(Goal)
    ->  run_case_result(Name, Expected, Actual, Indent, Failures)
    ;   write_indented_line(Indent, ['[FAIL] ', Name, ' -> goal failed']),
        Failures = 1
    ).

run_tests([], _, 0).
run_tests([Test|Tests], Indent, Failures) :-
    run_test(Test, Indent, FirstFailures),
    run_tests(Tests, Indent, RestFailures),
    Failures is FirstFailures + RestFailures.

run_case_result(Name, Expected, Actual, Indent, 0) :-
    Expected == Actual,
    write_indented_line(Indent, ['[PASS] ', Name]).
run_case_result(Name, Expected, Actual, Indent, 1) :-
    Expected \== Actual,
    term_string(Expected, ExpectedString),
    term_string(Actual, ActualString),
    write_indented_line(
        Indent,
        ['[FAIL] ', Name, ' -> expected ', ExpectedString, ', but got ', ActualString]
    ).

write_indented_line(Indent, Parts) :-
    format('~*c', [Indent, 0' ]),
    write_content(Parts),
    nl.

write_content(Parts) :-
    is_list(Parts),
    write_parts(Parts).
write_content(Part) :-
    \+ is_list(Part),
    write(Part).

write_parts([]).
write_parts([Part|Parts]) :-
    write(Part),
    write_parts(Parts).
